### PR TITLE
DAOS-5798 dfs: enhance support for reading symlinks

### PR DESCRIPTION
**Pull request by daltonbohning**
- Moved core logic of dfs_lookup into lookup_rel_path
- lookup_rel_path conditionally dereferences leaf symbolic
links and implements partial support for ".."
  - Recursion up to DFS_MAX_RECURSION = 40
- Changed dfuse calls to dfs_lookup_rel to pass O_NOFOLLOW
- Added dfs unit tests for dfs_lookup, dfs_lookup_rel, and the O_NOFOLLOW behavior.

Signed-off-by: Dalton Bohning <daltonx.bohning@intel.com>

### CODE REVIEW

**Comment by daltonbohning** [comment]
@mchaarawi @ashleypittman I'm working on this PR in conjunction with corresponding changes in mpifileutils. I'd appreciate any feedback on the current changes to make sure the changes to mpifileutils are valid.
For now, I have a new functional test introduced in this PR for convenience, but that can be separated into a separate PR if needed and merged after this PR.

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -1997,6 +2059,28 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
 	goto out;
 }
 
+int
+dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
+	   mode_t *mode, struct stat *stbuf)
+{
+	if (dfs == NULL || !dfs->mounted)
+	return EINVAL;
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c:2067]: missing a tab?

File: src/client/dfs/dfs.c [hunk 2]
```diff
@@ -1849,7 +1838,47 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
...
+lookup_rel_path_loop:
+
+		/*
+		 * Open the directory object one level up.
+		 * Since fetch_entry does not support ".",
+		 * we can't support ".." as the last entry,
+		 * nor can we support "../.." because we don't
+		 * have parent.parent_oid and parent.mode.
+		 * For now, represent this partial state with
+		 * parent_fully_valid.
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c:1850]: i have some thought on how to support .., but unfortunately it required a layout change and some modifications to rename. but something for the future

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -1849,7 +1838,47 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
...
+		 */
+		parent_fully_valid = true;
+		if (strcmp(token, "..") == 0) {
+			parent_fully_valid = false;
+
+			/* Cannot go outside the container */
+			if (daos_oid_cmp(parent.oid, dfs->root.oid) == 0) {
+				D_GOTO(err_obj, rc = ENOENT);
```
Comment (mchaarawi) [inline src/client/dfs/dfs.c:1858]: not required, but printing a debug message here might be helpful

**Review by mchaarawi — APPROVED** [review]

**Comment by mchaarawi** [comment]
im not sure if it's doing that, but all functional tests should run with this PR

**Comment by daltonbohning** [comment]
> im not sure if it's doing that, but all functional tests should run with this PR before you request landing

Right. I can't run the new functional tests yet, since the changes to mpifileutils have yet to be merged upstream, and thus also not in CI, so they would fail.

The goal is to:
- Get the DFS changes approved
- Merge the mpifileutils changes upstream
- Update mpifileutils in CI
- Land this PR

File: src/client/dfs/dfs.c (hunk 2 above)
Comment (daltonbohning) [reply src/client/dfs/dfs.c:1850]: Sure. It's an edge case anyway

File: src/client/dfs/dfs.c (hunk 3 above)
Comment (daltonbohning) [reply src/client/dfs/dfs.c:1881]: I agree. I added one

**Comment by daltonbohning** [comment]
I realized I completely forgot to check for infinite recursion. Added that, as well as some memory fixes.

**Comment by daltonbohning** [comment]
@mchaarawi 
1. I updated lookup_rel_path with a "depth" argument, to prevent infinite recursion. I put this as the last argument, even though it's only input, not output. My reasoning was that it's really more of a "meta" argument. Let me know if this isn't in line with our style.
2. I removed the functional tests due to some dependencies that aren't yet ready. They will be introduced separately.
3. I added unit tests to verify the new changes, as well as dfs_lookup[_rel] in general.

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfs/dfs.c [hunk 4]
```diff
@@ -3024,7 +3144,8 @@ dfs_access(dfs_t *dfs, dfs_obj_t *parent, const char *name, int mask)
 
 	D_ASSERT(entry.value);
 
-	rc = dfs_lookup(dfs, entry.value, O_RDONLY, &sym, NULL, NULL);
+	rc = lookup_rel_path(dfs, parent, entry.value, O_RDONLY, &sym,
```
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:3147]: Ah, yes, it does. I just added another call in the unit tests to verify this as well.

File: src/client/dfs/dfs.c [hunk 5]
```diff
@@ -1818,23 +1820,15 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
...
-	if (dfs == NULL || !dfs->mounted)
-		return EINVAL;
-	if (_obj == NULL)
-		return EINVAL;
-	if (path == NULL || strnlen(path, PATH_MAX) > PATH_MAX - 1)
-		return EINVAL;
-	if (path[0] != '/')
-		return EINVAL;
+	/* Arbitrarily stop recursion after 40 levels */
+	if (depth >= 40)
```
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:1826]: Agreed. Added, and a unit test for ELOOP.

File: src/client/dfs/dfs.c [hunk 6]
```diff
@@ -1818,23 +1820,15 @@ dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
...
 
-	/** if we added a prefix, check and skip over it */
-	if (dfs->prefix) {
-		if (strncmp(dfs->prefix, path, dfs->prefix_len) != 0)
-			return EINVAL;
-
-		path += dfs->prefix_len;
-	}
+	/* Only paths from root can be absolute */
+	if (path[0] == '/' && daos_oid_cmp(root->oid, dfs->root.oid) != 0)
```
[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/client/dfs/dfs.c:1830]: Yes, it is necessary. When `dfs_lookup` is called, the '/' is ALWAYS required (previous behavior).
But `lookup_rel_path` can be called recursively with a _**relative**_ path that does not start with '/', but for which the parent is the root. I couldn't think of a cleaner way than to have the duplicate check, other than to check right before making a recursive call, which would require a check at each place that calls `lookup_rel_path`.
I'm open to other suggestions, if you have any.

**Comment by daltonbohning** [comment]
Updated per comments and rebuilt against newer master.
@ashleypittman This doesn't seem to have broken anything in DFUSE, but please let me know if you see anything fishy.

File: src/client/dfs/dfs.c [hunk 7]
```diff
@@ -1805,9 +1809,10 @@ dfs_remove(dfs_t *dfs, dfs_obj_t *parent, const char *name, bool force,
 	return rc;
 }
 
-int
-dfs_lookup(dfs_t *dfs, const char *path, int flags, dfs_obj_t **_obj,
-	   mode_t *mode, struct stat *stbuf)
+static int
+lookup_rel_path(dfs_t *dfs, dfs_obj_t *root, const char *path, int flags,
+		dfs_obj_t **_obj, mode_t *mode, struct stat *stbuf,
+		size_t depth)
```
Comment (daltonbohning) [inline src/client/dfs/dfs.c:1815]: @ashleypittman This is the function I mentioned before. In order to handle symlink dereferencing, I needed to be able to lookup a path relative to some parent, which is what this function does. It's literally just `dfs_lookup`, but relative to some `root`.

**Review by mchaarawi — APPROVED** [review]

**Comment by mchaarawi** [comment]
you need a merge with master

**Comment by daltonbohning** [comment]
> you need a merge with master

@mchaarawi Just pushed. Thanks!

**Review by mchaarawi — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Comment by ashleypittman** [comment]
The NLT error is reported at https://build.hpdd.intel.com/job/daos-stack/job/daos/view/change-requests/job/PR-3970/14/valgrindResult/pid=39227,0x18/

File: src/client/dfs/dfs.c [hunk 8]
```diff
@@ -2255,15 +2358,33 @@ dfs_lookup_rel(dfs_t *dfs, dfs_obj_t *parent, const char *name, int flags,
...
-		/* Create a truncated version of the string */
-		D_STRNDUP(obj->value, entry.value, entry.value_len + 1);
-		if (obj->value == NULL)
-			D_GOTO(err_obj, rc = ENOMEM);
-		D_FREE(entry.value);
-		if (stbuf)
-			stbuf->st_size = entry.value_len;
+		if (flags & O_NOFOLLOW) {
+			/* Create a truncated version of the string */
+			D_STRNDUP(obj->value, entry.value, entry.value_len + 1);
```
Comment (ashleypittman) [inline src/client/dfs/dfs.c:2364]: Seems the segfault is because entry.value is NULL at this point.

Comment (daltonbohning) [reply src/client/dfs/dfs.c:2364]: Thanks @ashleypittman . It looks like I messed up the merge here. There should be a break statement before the S_IFLNK case. The S_IFREG is falling through here

Comment (ashleypittman) [reply src/client/dfs/dfs.c:2364]: I saw you'd retriggerd the Unit Test phase and I didn't think that was going to work.

Comment (daltonbohning) [reply src/client/dfs/dfs.c:2364]: For some reason, when I looked at the logs, I was seeing some erroneous error, so I just restarted. But I think maybe I had just clicked the wrong link :)

**Comment by daltonbohning** [comment]
@ashleypittman @mchaarawi Finally got this through CI... The only changes are really just the merge with master.

**Review by mchaarawi — APPROVED** [review]

**Review by ashleypittman — APPROVED** [review]

**Pull request merged**